### PR TITLE
Use session heartbeat to update CSRF token

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes https://github.com/nextcloud/server/issues/7155

Open questions/todo
- [x] Shall we allow anonymous users to request a token? Currently it's blocked because the @NoLoginRequired annotation hasn't been added.
- [x] Strict typing
- [x] Tests

cc @rullzer @nextcloud/security

### CODE REVIEW

**Comment by rullzer** [comment]
I would say unauthenticated users also should be able to get a token. Else you still get CSRF errors on public auth pages for example (login, password protected shares).

Getting a token is not that special or secret. So I see no reason to limit this.

File: core/Controller/CSRFTokenController.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
+<?php
+
```
Comment (MorrisJobke) [inline core/Controller/CSRFTokenController.php]: Strict types for new files? 🙈

File: core/Controller/CSRFTokenController.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+use OCP\AppFramework\Controller;
+use OCP\AppFramework\Http\JSONResponse;
+use OCP\IRequest;
+
+class CSRFTokenController extends Controller {
+
+	/** @var CsrfTokenManager */
+	private $tokenManager;
+
+	public function __construct($appName, IRequest $request,
```
Comment (rullzer) [inline core/Controller/CSRFTokenController.php]: `string $appName`

File: core/Controller/CSRFTokenController.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		parent::__construct($appName, $request);
+		$this->tokenManager = $tokenManager;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 * @return JSONResponse
+	 */
+	public function index() {
```
Comment (rullzer) [inline core/Controller/CSRFTokenController.php]: `: JSONResponse`

File: core/Controller/CSRFTokenController.php (hunk 1 above)
Comment (ChristophWurst) [reply core/Controller/CSRFTokenController.php]: That would make a lot of sense :see_no_evil:

**Review by rullzer — DISMISSED** [review]

**Comment by ChristophWurst** [comment]
Will squash my commits once I know this gets approved :+1:

File: core/Controller/CSRFTokenController.php [hunk 4]
```diff
@@ -34,7 +35,12 @@ class CSRFTokenController extends Controller {
 	/** @var CsrfTokenManager */
 	private $tokenManager;
 
-	public function __construct($appName, IRequest $request,
+	/**
+	 * @param string $appName
+	 * @param IRequest $request
+	 * @param CsrfTokenManager $tokenManager
+	 */
```
Comment (MorrisJobke) [inline core/Controller/CSRFTokenController.php]: Just a little sidenote: This kind of comments makes absolutely no sense at all and can be dropped in my opinion. All of this information is already present in the type hinting and thus we could make it a bit less cluttered to not repeat useless stuff. This does not block this PR at all ;)

@nickvergessen @rullzer @blizzz Opinions?

Comment (ChristophWurst) [reply core/Controller/CSRFTokenController.php]: Sure, totally fine by me

Comment (blizzz) [reply core/Controller/CSRFTokenController.php]: Lemme quote The Offspring on this: 

> Yeah, I'm a pyro
I wanna burn it up
Yeah, I'm a pyro

:fire: 

so as long as stuff does not need to be backported. In this case we're good. :fire:

**Comment by rullzer** [comment]
Please also cleanup https://github.com/nextcloud/server/blob/fb890807c0c17549566690512549c8d4ae5813e3/core/routes.php#L153

File: core/Controller/CSRFTokenController.php (hunk 4 above)
Comment (rullzer) [reply core/Controller/CSRFTokenController.php]: Yes was my idea as well we can kill this now. And only have it in places where it makes sense.

**Review by rullzer — APPROVED** [review]
awesome stuff!

**Review by blizzz — APPROVED** [review]

**Comment by blizzz** [comment]
@ChristophWurst squash it :)

**Comment by ChristophWurst** [comment]
> Please also cleanup

Done :wink: 

> @ChristophWurst squash it :)

Done :wink:

**Pull request merged**